### ISSUE TITLE
command "palette"

### ISSUE THREAD

**Issue by teto** [neovim/neovim#6240]
One feature that was discussed several times on gitter and would really help beginners with vim would be some kind of "palette" like in sublimetext:
![Sublime palette](http://docs.sublimetext.info/en/latest/_images/command-palette-0001.png)

If I take my experience with vim (I believe similar to many others), I was a bit put off by simple things I wanted to do but had to lookup again & again because I had forgotten or didn't know the exact name of the setting. For instance, to configure tabs vs spaces: simple actions like 'retab' or settings ('tabstop', 'softtabstop', etc).
Sure there is the help, but it's fragmented over different files and you have first to learn how it works, which is not quite easy (with my locale, I have to hold 3 keys to jump to a tag).
That's where a palette comes handy, i.e. you type whant you want to do and it filters the different possible actions.
Such a palette could come handy to search things as https://github.com/neovim/neovim/pull/6236 (get_keymap API).
A  [plugin already exists for CtrlP](https://github.com/dbeecham/ctrlp-commandpalette.vim) but it's quite limited as it's fed manually. Neovim has this interesting file "/src/nvim/options.lua" that lists all settings and allows to automate things. One problem is it lacks proper "English" descriptions for the settings.
Just for fun, I made a PoC to configure neovim from a GUI (more for fun than anything else) along with this script (my first haskell program https://github.com/teto/vim-config/blob/master/test.hs) to extract oneliner descriptions from doc/quickref.txt.
Now I wonder what would be the best way to provide this feature with internationalization (gettext through cmake/RunXgettext.cmake) and across guis: a plugin, menus etc ?

Right now I use menus as a kind of palette:
```
menu Spell.EN_US :setlocal spell spelllang=en_us \| call histadd('cmd', 'setlocal spell spelllang=en_us')<CR>
menu Search.CurrentBuffer :exe Grepper -grepprg rg --vimgrep $* $.
menu Tabs.S2 :set  tabstop=2 softtabstop=2 sw=2<CR>
menu Tabs.S4 :set ts=4 sts=4 sw=4<CR>
```
I believe we could have a selection widget (like wildmenu https://github.com/neovim/neovim/pull/6168), that could be ofc externalized in which we could feed different items { description: command } like "Search in current buffer" : ":exe Grepper -grepprg rg --vimgrep $* $.".
Then menus can be generated from options.lua.

I believe we could add a  short_desc fields in options.lua entries, have a plugin generate mappings between description <-> commands and send items through a selection widget (wildmenu-like).

My concern is about internationalization of such menu items. If I take my locale, every gettext string is translated but I doubt that's the case for menus or the help (can't get the help in French at least on Ubuntu). The gettext workflow/process seems more robust. 

(I leave this as a reference for myself: http://vim.wikia.com/wiki/Vim_goes_INTL_-_Translating_Messages,_Menus,_Encodings)

**Comment by justinmk** [comment]
> what would be the best way to provide this feature with internationalization (gettext through cmake/RunXgettext.cmake) and across guis: a plugin, menus etc ?

You mean a built in fuzzy finder? 

src/nvim/options.lua is compiled so it can use gettext like any other C code.

**Comment by tjdevries** [comment]
This is basically the end goal of my Descriptive Maps feature that I requested / am working on.

I think if we provide an API to get these options, people should be able to integrate it into their workflow as they desire. Some might prefer Denite, some might like fzf.

I think neovim should provide `nvim_get_quickref`, `nvim_get_keymap`, ... where all of them return a list of dictionaries with a 'description' key or something like that.

**Comment by teto** [comment]
> You mean a built in fuzzy finder?

I suppose in the end, it comes down to this + having UI widgets accessible from plugins. I do like fzf but if it could open its selector with a UI widget, that would be better.
I chose to raise the issue now because I see similitudes between such a widget and wildmenu externalization PR.

**Comment by justinmk** [comment]
> I do like fzf but if it could open its selector with a UI widget, that would be better.

Why?

**Comment by fmoralesc** [comment]
> Why?

A composite widget/buffer that would react to changes to some input string would be nice to have. Last time I checked (when I built my vim-pad plugin) you had to do some juggling to implement something like unite, fzf et al, so some kind of generic facility for this greatly simplify this, specially for cases where the behaviour of the buffer is intended to be very specialized (unite and fzf are nice, but they impose limitations and are difficult to customize deeply without touching the core functionality). However, implementing the idea would need some careful planning, because it might require introducing some new type of buffer with quite a distinct behaviour (off the top of my head: all changes would be redirected to an input buffer, with a slave buffer providing a view without any mappings other than basic movements).

So, generically, one could implement something like this with (just some thoughts, I'm not proposing this API)

``` vim
function initialize()
" set prompt, maybe add highlighting to <picker> buffer
endfunction

function populate(picker)
    call pickerclear(a:picker)
    call pickeradd(a:picker, [...])
endfunction

function launch_picker(init, event)
    vsplit
    set buftype=picker
    " the slave buffer is created and assigned an id, kept as <picker>
    execute 'call '.a:init'.'(<picker>)'
    execute 'au <buffer> TextChanged '.a:event('<picker>')
endfunction

call launch_picker('initialize', 'populate')
```

**Comment by teto** [comment]
>> I do like fzf but if it could open its selector with a UI widget, that would be better.
>
> Why?

because I want a nice looking vim :D ! When using nvim-qt it feels the same as using the terminal, I would hope for sthg fancier.
As a first step, I will try to propose an fzf based palette. I am only blocked by some python-client problems but I've added the short descriptions to options.lua (https://github.com/teto/vim-config/blob/master/options.lua) as well as a script that converts it to csv. From there, it will be easy to generate fzf entries.

There might be future usecases where the UI might want to be responsible for the input and just return a result rather than - as in the current wildmenu PR - just reflects what vim sends.

**Comment by justinmk** [comment]
> where the UI might want to be responsible for the input and just return a result rather than - as in the current wildmenu PR - just reflects what vim sends.

I don't follow. If it doesn't reflect what Vim sends then how is it any different than a separate plugin?

**Comment by teto** [comment]
I was thinking of things similar to what was asked by lunixbochs on gitter:

>is there a way for me to select an item in the popup menu (like if someone clicks in the ui)?
>I only see the events to render it, nothing to interact with it

like the UI could just return a result and let nvim close the pum for instance.

I've made a poc for a palette, seems like the descriptions could use some improvement.
https://github.com/teto/nvim-palette

I will prepare a PR with changes to options.lua.

**Comment by bfredl** [comment]
> is there a way for me to select an item in the popup menu (like if someone clicks in the ui)?
I only see the events to render it, nothing to interact with it

I have some old WIP code for this https://github.com/neovim/neovim/compare/master...bfredl:pum_select4?expand=1

I just rebased it, but I haven't checked if it still working. I haven't really time to work on it right now, but if someone wants to pick it up...

**Comment by teto** [comment]
I've asked here https://github.com/vim/vim/issues/1563 if vim could provide a way to retrieve menu options.

Pr witch changes to options.lua https://github.com/neovim/neovim/pull/6288

**Comment by justinmk** [comment]
@teto Why JSON instead of a VimL Dict?

**Comment by teto** [comment]
vimL dict works too. I suggested json because of python gettext support but with json_encode/decode vimL_dict might be a better choice as it might be easier to generate.

**Comment by dzhou121** [comment]
Is it possible for other plugins to "register" their commands to this command palette. I think that would be handy.

**Comment by justinmk** [comment]
`nvim_get_keymap()` was implemented, and the discussion here doesn't have more specifics, so let's close this and track specific tasks in new issues. https://github.com/neovim/neovim/issues/7833 is a request for `nvim_get_commands()`, I'll leave it to others to add more requests.

**Comment by teto** [comment]
telescope.nvim is not quite there yet but it's promising https://github.com/nvim-lua/telescope.nvim/pull/133
